File: packages/protocol/test/messages.test.ts
import { describe, expect, it } from "vitest";
import {
  FrameType,
  MAX_FRAME_PAYLOAD_BYTES,
  type Message,
  ProtocolError,
  ZERO_REQUEST_ID,
  decodeMessage,
  encodeFrame,
  encodeMessage,
  newRequestId,
} from "../src/index.js";

function roundtrip(msg: Message): Message {
  return decodeMessage(encodeMessage(msg));
}

describe("encodeMessage / decodeMessage roundtrip", () => {
  it("hello", () => {
    const msg: Message = {
      kind: "hello",
      requestId: ZERO_REQUEST_ID,
      tunnelId: "t1",
      agentVersion: "1.0.0",
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("helloAck", () => {
    const msg: Message = {
      kind: "helloAck",
      requestId: ZERO_REQUEST_ID,
      tunnelId: "t1",
      publicUrl: "https://t1.example.dev",
      heartbeatIntervalMs: 20_000,
      heartbeatTimeoutMs: 60_000,
      requestTimeoutMs: 30_000,
      maxPayloadBytes: MAX_FRAME_PAYLOAD_BYTES,
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("requestStart", () => {
    const msg: Message = {
      kind: "requestStart",
      requestId: newRequestId(),
      method: "POST",
      path: "/api/things?x=1",
      headers: [
        ["Content-Type", "application/json"],
        ["X-Dup", "1"],
        ["X-Dup", "2"],
      ],
      hasBody: true,
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("requestBody", () => {
    const msg: Message = {
      kind: "requestBody",
      requestId: newRequestId(),
      data: new Uint8Array([1, 2, 3, 4]),
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("requestEnd", () => {
    const msg: Message = { kind: "requestEnd", requestId: newRequestId() };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("responseStart", () => {
    const msg: Message = {
      kind: "responseStart",
      requestId: newRequestId(),
      status: 404,
      headers: [
        ["Set-Cookie", "a=1"],
        ["Set-Cookie", "b=2"],
      ],
      hasBody: false,
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("responseBody", () => {
    const msg: Message = {
      kind: "responseBody",
      requestId: newRequestId(),
      data: new Uint8Array([9, 8, 7]),
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("responseEnd", () => {
    const msg: Message = { kind: "responseEnd", requestId: newRequestId() };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it.each(["timeout", "client_disconnected", "upstream_error", "shutdown"] as const)(
    "cancel (%s)",
    (reason) => {
      const msg: Message = { kind: "cancel", requestId: newRequestId(), reason };
      expect(roundtrip(msg)).toEqual(msg);
    },
  );

  it("ping", () => {
    const msg: Message = { kind: "ping" };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("pong", () => {
    const msg: Message = { kind: "pong" };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("error", () => {
    const msg: Message = {
      kind: "error",
      requestId: newRequestId(),
      code: "upstream_unreachable",
      message: "connection refused",
    };
    expect(roundtrip(msg)).toEqual(msg);
  });

  it("ping/pong always encode a zero request id", () => {
    const pingEncoded = encodeMessage({ kind: "ping" });
    expect(pingEncoded.slice(2, 18)).toEqual(ZERO_REQUEST_ID);
    const pongEncoded = encodeMessage({ kind: "pong" });
    expect(pongEncoded.slice(2, 18)).toEqual(ZERO_REQUEST_ID);
  });
});

describe("oversize payload handling", () => {
  it("rejects encodeMessage of a requestBody exceeding the max payload", () => {
    const msg: Message = {
      kind: "requestBody",
      requestId: newRequestId(),
      data: new Uint8Array(MAX_FRAME_PAYLOAD_BYTES + 1),
    };
    expect(() => encodeMessage(msg)).toThrowError(ProtocolError);
  });

  it("rejects decodeMessage of a frame whose declared length exceeds the max", () => {
    const header = new Uint8Array(22);
    header[0] = 1;
    header[1] = FrameType.RequestBody;
    const view = new DataView(header.buffer);
    view.setUint32(18, MAX_FRAME_PAYLOAD_BYTES + 1, false);
    expect(() => decodeMessage(header)).toThrowError(ProtocolError);
  });
});

describe("malformed JSON payloads", () => {
  it("throws invalid_json for non-JSON bytes", () => {
    const encoded = encodeFrame(
      FrameType.Hello,
      ZERO_REQUEST_ID,
      new TextEncoder().encode("not json"),
    );
    try {
      decodeMessage(encoded);
      expect.unreachable();
    } catch (err) {
      expect(err).toBeInstanceOf(ProtocolError);
      expect((err as ProtocolError).code).toBe("invalid_json");
    }
  });

  it("throws invalid_json for JSON that is not an object", () => {
    const encoded = encodeFrame(
      FrameType.Hello,
      ZERO_REQUEST_ID,
      new TextEncoder().encode("[1,2,3]"),
    );
    expect(() => decodeMessage(encoded)).toThrow(ProtocolError);
  });

  it("throws invalid_json when a required field is missing", () => {
    const encoded = encodeFrame(
      FrameType.Hello,
      ZERO_REQUEST_ID,
      new TextEncoder().encode('{"tunnelId":"t1"}'),
    );
    try {
      decodeMessage(encoded);
      expect.unreachable();
    } catch (err) {
      expect((err as ProtocolError).code).toBe("invalid_json");
    }
  });

  it("throws invalid_json when a field has the wrong type", () => {
    const encoded = encodeFrame(
      FrameType.RequestStart,
      newRequestId(),
      new TextEncoder().encode('{"method":"GET","path":"/","headers":[],"hasBody":"yes"}'),
    );
    try {
      decodeMessage(encoded);
      expect.unreachable();
    } catch (err) {
      expect((err as ProtocolError).code).toBe("invalid_json");
    }
  });

  it("throws invalid_json when headers are malformed", () => {
    const encoded = encodeFrame(
      FrameType.RequestStart,
      newRequestId(),
      new TextEncoder().encode(
        '{"method":"GET","path":"/","headers":[["only-one-item"]],"hasBody":false}',
      ),
    );
    expect(() => decodeMessage(encoded)).toThrow(ProtocolError);
  });

  it("throws invalid_json for an invalid cancel reason", () => {
    const encoded = encodeFrame(
      FrameType.Cancel,
      newRequestId(),
      new TextEncoder().encode('{"reason":"bogus"}'),
    );
    expect(() => decodeMessage(encoded)).toThrow(ProtocolError);
  });

  it("throws invalid_json for an invalid error code", () => {
    const encoded = encodeFrame(
      FrameType.Error,
      newRequestId(),
      new TextEncoder().encode('{"code":"bogus","message":"x"}'),
    );
    expect(() => decodeMessage(encoded)).toThrow(ProtocolError);
  });
});

File: apps/cli/internal/protocol/messages_test.go
package protocol

import (
	"encoding/binary"
	"errors"
	"reflect"
	"testing"
)

func roundtrip(t *testing.T, m Message) Message {
	t.Helper()
	encoded, err := EncodeMessage(m)
	if err != nil {
		t.Fatalf("EncodeMessage returned error: %v", err)
	}
	decoded, err := DecodeMessage(encoded)
	if err != nil {
		t.Fatalf("DecodeMessage returned error: %v", err)
	}
	return decoded
}

func assertRoundtrip(t *testing.T, m Message) {
	t.Helper()
	got := roundtrip(t, m)
	if !reflect.DeepEqual(got, m) {
		t.Errorf("roundtrip mismatch:\n got:  %#v\n want: %#v", got, m)
	}
}

func TestMessageRoundtrips(t *testing.T) {
	id := NewRequestID()

	t.Run("Hello", func(t *testing.T) {
		assertRoundtrip(t, Hello{RequestID: ZeroRequestID, TunnelID: "t1", AgentVersion: "1.0.0"})
	})

	t.Run("HelloAck", func(t *testing.T) {
		assertRoundtrip(t, HelloAck{
			RequestID:           ZeroRequestID,
			TunnelID:            "t1",
			PublicURL:           "https://t1.example.dev",
			HeartbeatIntervalMs: 20000,
			HeartbeatTimeoutMs:  60000,
			RequestTimeoutMs:    30000,
			MaxPayloadBytes:     MaxPayloadBytes,
		})
	})

	t.Run("RequestStart", func(t *testing.T) {
		assertRoundtrip(t, RequestStart{
			RequestID: id,
			Method:    "POST",
			Path:      "/api/things?x=1",
			Headers: HeaderPairs{
				{"Content-Type", "application/json"},
				{"X-Dup", "1"},
				{"X-Dup", "2"},
			},
			HasBody: true,
		})
	})

	t.Run("RequestBody", func(t *testing.T) {
		assertRoundtrip(t, RequestBody{RequestID: id, Data: []byte{1, 2, 3, 4}})
	})

	t.Run("RequestEnd", func(t *testing.T) {
		assertRoundtrip(t, RequestEnd{RequestID: id})
	})

	t.Run("ResponseStart", func(t *testing.T) {
		assertRoundtrip(t, ResponseStart{
			RequestID: id,
			Status:    404,
			Headers: HeaderPairs{
				{"Set-Cookie", "a=1"},
				{"Set-Cookie", "b=2"},
			},
			HasBody: false,
		})
	})

	t.Run("ResponseBody", func(t *testing.T) {
		assertRoundtrip(t, ResponseBody{RequestID: id, Data: []byte{9, 8, 7}})
	})

	t.Run("ResponseEnd", func(t *testing.T) {
		assertRoundtrip(t, ResponseEnd{RequestID: id})
	})

	for _, reason := range []CancelReason{CancelTimeout, CancelClientDisconnected, CancelUpstreamError, CancelShutdown} {
		t.Run("Cancel_"+string(reason), func(t *testing.T) {
			assertRoundtrip(t, Cancel{RequestID: id, Reason: reason})
		})
	}

	t.Run("Ping", func(t *testing.T) {
		assertRoundtrip(t, Ping{})
	})

	t.Run("Pong", func(t *testing.T) {
		assertRoundtrip(t, Pong{})
	})

	t.Run("ErrorMessage", func(t *testing.T) {
		assertRoundtrip(t, ErrorMessage{RequestID: id, Code: ErrorCodeUpstreamUnreachable, Message: "connection refused"})
	})
}

func TestPingPongAlwaysEncodeZeroRequestID(t *testing.T) {
	encoded, err := EncodeMessage(Ping{})
	if err != nil {
		t.Fatalf("EncodeMessage returned error: %v", err)
	}
	var requestID RequestID
	copy(requestID[:], encoded[2:18])
	if requestID != ZeroRequestID {
		t.Errorf("Ping request id = %x, want zero", requestID)
	}

	encoded, err = EncodeMessage(Pong{})
	if err != nil {
		t.Fatalf("EncodeMessage returned error: %v", err)
	}
	copy(requestID[:], encoded[2:18])
	if requestID != ZeroRequestID {
		t.Errorf("Pong request id = %x, want zero", requestID)
	}
}

func TestEncodeMessageRejectsOversizeBody(t *testing.T) {
	msg := RequestBody{RequestID: NewRequestID(), Data: make([]byte, MaxPayloadBytes+1)}
	_, err := EncodeMessage(msg)
	if !errors.Is(err, ErrPayloadTooLarge) {
		t.Errorf("errors.Is(err, ErrPayloadTooLarge) = false, want true (err=%v)", err)
	}
}

func TestDecodeMessageRejectsOversizeDeclaredLength(t *testing.T) {
	header := make([]byte, HeaderSize)
	header[0] = Version
	header[1] = byte(FrameRequestBody)
	binary.BigEndian.PutUint32(header[18:22], uint32(MaxPayloadBytes+1))
	_, err := DecodeMessage(header)
	if !errors.Is(err, ErrPayloadTooLarge) {
		t.Errorf("errors.Is(err, ErrPayloadTooLarge) = false, want true (err=%v)", err)
	}
}

func TestDecodeMessageMalformedJSON(t *testing.T) {
	t.Run("non-JSON bytes", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{Type: FrameHello, RequestID: ZeroRequestID, Payload: []byte("not json")})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})

	t.Run("JSON that is not an object", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{Type: FrameHello, RequestID: ZeroRequestID, Payload: []byte("[1,2,3]")})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})

	t.Run("missing required field", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{Type: FrameHello, RequestID: ZeroRequestID, Payload: []byte(`{"tunnelId":"t1"}`)})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})

	t.Run("wrong field type", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{
			Type:      FrameRequestStart,
			RequestID: NewRequestID(),
			Payload:   []byte(`{"method":"GET","path":"/","headers":[],"hasBody":"yes"}`),
		})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})

	t.Run("malformed header pair", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{
			Type:      FrameRequestStart,
			RequestID: NewRequestID(),
			Payload:   []byte(`{"method":"GET","path":"/","headers":[["only-one-item"]],"hasBody":false}`),
		})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})

	t.Run("invalid cancel reason", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{Type: FrameCancel, RequestID: NewRequestID(), Payload: []byte(`{"reason":"bogus"}`)})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})

	t.Run("invalid error code", func(t *testing.T) {
		encoded, err := EncodeFrame(Frame{
			Type:      FrameError,
			RequestID: NewRequestID(),
			Payload:   []byte(`{"code":"bogus","message":"x"}`),
		})
		if err != nil {
			t.Fatalf("EncodeFrame returned error: %v", err)
		}
		_, err = DecodeMessage(encoded)
		if !errors.Is(err, ErrInvalidJSON) {
			t.Errorf("errors.Is(err, ErrInvalidJSON) = false, want true (err=%v)", err)
		}
	})
}
